:slock modifications

ezekieltem opened the pull request:
# New Command ":rig"
Command: `:rig <Player(s)> <RigType>`
Changes the provided player(s)' rigtype

# ":slock" Changes
New Command `:slock <Active> <AdminLevel>`
The change allows for admin's to lock the server to a minimum admin level.
The provided admin level cannot be higher than the executor's own, and the executor must have a higher or equal admin level to the current server lock level to change it.
Every time a player disconnects if there is no player with an admin level higher than or equal to the server lock level, the server lock is disabled.

ezekieltem committed:
Add support for minimum AdminLevel in a ServerLock

ezekieltem committed:
Create rig.lua

Add a command to change the RigType of given players

ezekieltem committed:
Account for new rig command

TheNexusAvenger commented:
Allowing to specify an admin level for `slock` sounds fine, but I am interested in the use case. `rig` I would like to know the use case because changing rig types to R6 on R15 games and R15 to R6 games tends to cause a lot of things to break. I'm hesitant to say approve a new command that will break a lot of games if used.

ezekieltem commented:
Specifying of an admin level for `slock` would be useful in cases where the default admin level is 0 or higher.
For `rig` a possible use case would be testing a rig type for a game that was built and testing with a separate rig type. I've just added it every time I've used NexusAdmin and felt others could find use for it.

TheNexusAvenger commented:
`rig` feels too destructive for me to roll out. While testing is valid, the general direction is to either lock the game to R6 or to R15. Managing both can be a massive pain, as I have learned with Ultimate Boxing. In the games I have maintained or maintain with Nexus Admin (Innovation Security Training Facility, Innovation Inc Thermal Power Plant, Tsunami Game, Survive a Plane Crash, and all my smaller games), *all* of them would dump nonsense errors to the output and Roblox's Error Report.

`slock` I can consider accepting, but it may not fit the use case you desire. If you have a non-default admin level (several parts of the system assume 0 is the lowest admin level), *every* run of `slock` will need to include this non-default admin level. For that type of use case, it may be time to add a generic configuration to the loader.

ezekieltem commented:
For `rig` that is a fair point, I'm just going to take it and have it as a command that can just be put under the loader.

For `slock` the default server lock level a possible solution that wouldn't require the addition of any configuration would be adding 1 to the default admin level specified in the configuration.

ezekieltem committed:
Remove “rig” from loader

ezekieltem committed:
Remove rig module

ezekieltem committed:
Set default slock level to 1 above default admin level

TheNexusAvenger commented on src/IncludedCommands/Administrative/slock.lua:
```diff
@@ -13,44 +13,89 @@ local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
+                return "Cannot set ServerLock above your own AdminLevel."
+            end
+            Api.CommandData.ServerLock.Locked = not Api.CommandData.ServerLock.Locked
+            Api.CommandData.ServerLock.AdminLevel = Api.Configuration.DefaultAdminLevel + 1
```
`Api.Configuration.DefaultAdminLevel + 1` is *not* a safe assumption to go with. `DefaultAdminLevel` is the admin level given to players by default (typically <0), not the minimum level to be considered an admin (currently 0). At least 1 game I have worked on used multiple admin levels below 0 for player-facing functionality. For when the value is changed, this would make `slock` not default to 0.

TheNexusAvenger reviewed (commented):
I think the changes to `Api.CommandData.ServerLocked` can be simplifed from a new table to just a `number?`. If it is locked, there will be some number. If it is not locked, it will be nil.

ezekieltem committed:
Update slock.lua

Changed Api.CommandData.ServerLocked to either be a number or nil; Account for Api.Configuration.DefaultServerLockAdminLevel

ezekieltem committed:
Update Configuration.lua

Add DefaultServerLockAdminLevel

ezekieltem committed:
Update Loader.server.lua

Add DefaultServerLockAdminLevel

ezekieltem committed:
Update slock.lua

Change DefaultServerLockLevel to DefaultServerLockAdminLevel

ezekieltem commented:
I'm hoping I both understood what you meant, and that I didn't make any mistakes with the code changes. I am having to use Github Mobile for this as of the current moment.

ezekieltem committed:
Update Types.lua

Add `DefaultServerLockAdminLevel: number,` to account for new configurations

ezekieltem committed:
Update api.md

Add `Configuration.DefaultServerLockAdminLevel`

TheNexusAvenger commented on Loader.server.lua:
```diff
@@ -22,6 +22,9 @@ local Config = {
     DefaultAdminLevel = -1,
 
+    --Default Admin Level for Server Lock (slock)
+    DefaultServerLockAdminLevel = 0,
```
While I like this being configurable, I don't like it being at the top-level. More may come in the future and make the configuration very long. This is something I would be fine with (feel free to suggest an alternative name).

```lua
CommandConfigurations = { --CommandSettings, CommandOverrides, or something else would be fine.
    DefaultServerLockAdminLevel = 0,
},
```

With that being said, I am aware that it makes the logic in `Configuration.lua` a bit tricky, since you have to correct this being `nil` *and* `DefaultServerLockAdminLevel` being `nil`.

ezekieltem replied:
I had initially put it there since the `DefaultAdminLevel` and `Admins` Configuration is right next to it, and didn't know where else to put it.

For `Configuration.lua` a possible way to address this (based on what I initially thought of) would be having a constant at the top of the script that is the defaults for `CommandConfigurations` and run a loop on the reference settings and compare to the configuration settings, and make changes accordingly.

```lua
local COMMAND_CONFIGURATIONS = {
    --Default Admin Level for Server Lock (slock).
    DefaultServerLockAdminLevel = 0,

    --Other future settings accordingly.
}
```

Commenting this before actually writing any code to do this, as you (having a lot more experience than I do) may know of a more effective and efficient way.

TheNexusAvenger replied:
I didn't have the idea of putting the defaults at the top, but I like it. Only thing I'd change is to name it `DEFAULT_COMMAND_CONFIGURATIONS`.

TheNexusAvenger commented on src/IncludedCommands/Administrative/slock.lua:
```diff
@@ -13,44 +13,92 @@ local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
     },
     ServerLoad = function(Api: Types.NexusAdminApiServer)
-        Api.CommandData.ServerLocked = false
+        Api.CommandData.ServerLocked = nil
```
While it was here before, I'd like the `ServerLocked` variable renamed to something like `MinimumAdminLevel`, `CurrentMinimumAdminLevel`, `RequiredAdminLevel`, or something else that makes it clear it is an admin level instead of a bool.

TheNexusAvenger commented on src/IncludedCommands/Administrative/slock.lua:
```diff
@@ -13,44 +13,92 @@ local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
+                return
+            end
+
+            -- Set HighestAdminLevel to an unreasonably low number to avoid error
+            local HighestAdminLevel = -math.huge
+            for _, Player in Players:GetPlayers() do
+                local AdminLevel = Api.Authorization:YieldForAdminLevel(Player)
+                if AdminLevel > HighestAdminLevel then
+                    HighestAdminLevel = AdminLevel
+                end
+            end
+            
+            -- If there is no player with a higher AdminLevel than the current ServerLock AdminLevel disable ServerLock
```
*Very* minor nitpick: the comments should end with punctuation and have no leading spaces.

TheNexusAvenger commented on src/IncludedCommands/Administrative/slock.lua:
```diff
@@ -13,44 +13,92 @@ local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
         local Util = IncludedCommandUtil.ForContext(CommandContext)
         local Api = Util:GetApi()
+        
+        if Api.Authorization:YieldForAdminLevel(CommandContext.Executor) < (Api.CommandData.ServerLocked or -math.huge) then
+            return "Cannot change ServerLock above your own AdminLevel."
+        end
```
I don't like the `or -math.huge` part of this statement. I'd prefer a `nil` check instead.
```lua
if Api.CommandData.ServerLocked and Api.Authorization:YieldForAdminLevel(CommandContext.Executor) < Api.CommandData.ServerLocked then
    return "Cannot change ServerLock above your own AdminLevel."
end
```

TheNexusAvenger reviewed (commented):
Sorry for the delay. I've been distracted with other things totally not related to LEGO Universe.

Note: I still haven't verified the functionality, but it seems fine. All cases I would think of accounting for are covered. Below is going through every detail I can think of, including nitpicks.

TheNexusAvenger commented:
I've also removed `:rig` from the title to make it easier to understand what this pull request does from the title.

ezekieltem committed:
Update slock.lua

Renamed "ServerLocked" to "MinimumAdminLevel"

Added a variable definition for `Api.Configuration.CommandConfigurations.DefaultServerLockAdminLevel` to avoid repetitive calls

Added a AdminLevel check in an overlooked place that could have resulted in a player setting the MinimumAdminLevel higher than their own AdminLevel

ezekieltem committed:
Update Configuration.lua

Added `DEFAULT_COMMAND_CONFIGURATIONS` to the top of the script.

Added new Configuration `CommandConfigurations`

ezekieltem committed:
Update Loader.server.lua

Added configuration `CommandConfigurations`

Moved `DefaultServerLockAdminLevel` into `CommandConfigurations`

ezekieltem commented:
I believe I covered everything, please call out anything I may have missed.

TheNexusAvenger commented on docs/api.md:
```diff
@@ -83,6 +83,9 @@ Keys used to open the command bar.
 Default admin level that will be given to players
 when the join.
 
+#### `Configuration.DefaultServerLockAdminLevel: number`
+Default admin level to use in "slock" when a AdminLevel is not provided.
```
Changing this was missed. There is now `CommandConfigurations` that contains this.

TheNexusAvenger commented on src/Types.lua:
```diff
@@ -203,6 +203,7 @@ export type Configuration = {
     CommandPrefix: string,
     ActivationKeys: {Enum.KeyCode},
     DefaultAdminLevel: number,
+    DefaultServerLockAdminLevel: number,
```
This type is out of date. Suggested version:
```lua
CommandConfigurations: {
    DefaultServerLockAdminLevel: number,
    [string]: any,
},
```

TheNexusAvenger commented on src/Common/Configuration.lua:
```diff
@@ -48,6 +53,12 @@ function Configuration.new(ConfigurationTable: {[string]: any}): Types.Configura
     self.FeatureFlagOverrides = ConfigurationTable.FeatureFlagOverrides or {}
+    self.CommandConfigurations = ConfigurationTable.CommandConfigurations or DEFAULT_COMMAND_CONFIGURATIONS
+    
+    -- Confirm all DEFAULT_COMMAND_CONFIGURATIONS exist within CommandConfigurations
```
Same nitpick as before: should end in punctuation and no leading space.

TheNexusAvenger commented on src/Common/Configuration.lua:
```diff
@@ -48,6 +53,12 @@ function Configuration.new(ConfigurationTable: {[string]: any}): Types.Configura
+    
+    -- Confirm all DEFAULT_COMMAND_CONFIGURATIONS exist within CommandConfigurations
+    for Name, Value in DEFAULT_COMMAND_CONFIGURATIONS do
+        self.CommandConfigurations[Name] = self.CommandConfigurations[Name] or Value
```
Right now, this would work, but if a value came around that was a bool, this would break. A default of `true` when the user has it set to `false` would override it back to `true`. Suggestion:

```lua
for Name, Value in DEFAULT_COMMAND_CONFIGURATIONS do
    if self.CommandConfigurations[Name] ~= nil then continue end
    self.CommandConfigurations[Name] = self.CommandConfigurations[Name]
end
```

When you make this mistake in a game, you won't forget it. Ultimate Boxing's feature flags had this bug, which meant turning off buggy features just didn't work for a while. That was like 8 years ago.

TheNexusAvenger reviewed (commented):
Getting very close. It is down to 1 concern, 1 nitpick, and 2 things not updated to the latest commit.

ezekieltem committed:
Update slock.lua

Fixed *another* oversight in allowing a player to set the admin level above their own if `Api.Configuration.CommandConfigurations.DefaultServerLockAdminLevel` is above their own admin level

ezekieltem committed:
Update Configuration.lua

Account for future CommandConfigurations that have a Bool value

ezekieltem committed:
nitpick fix

ezekieltem committed:
nitpick fix... again... (how blind can I be)

ezekieltem committed:
Update api.md

Correct docs to reflect recent changes.

ezekieltem committed:
Update Types.lua

Correct `Configuration.CommandConfigurations` typing

ezekieltem commented:
Believe I got everything, if I didn't call me out on it so I can label myself as blind, and then fix it.

TheNexusAvenger reviewed (approved):
I'm fine with merging it as-is, but there is 1 more change to the logic I want to make to *try* to reduce how many if conditions there are. If you are fine with me handling instead of asking you to do it, I'll merge the changes as-is.

ezekieltem commented:
Go for it, I have been out of programming for a bit and it would likely just extend this pull request longer than needed.
